src/main/java/mate/project/repository/book/BookSpecificationBuilder.java

```diff
@@ -0,0 +1,45 @@
+package mate.project.repository.book;
+
+import java.math.BigDecimal;
+import lombok.RequiredArgsConstructor;
+import mate.project.dto.BookSearchParameters;
+import mate.project.model.Book;
+import mate.project.repository.SpecificationBuilder;
+import mate.project.repository.SpecificationProviderManager;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> spec = Specification.where(null);
+        if (searchParameters.titles() != null && searchParameters.titles().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("title")
+                    .getSpecification(searchParameters.titles()));
+        }
+        if (searchParameters.authors() != null && searchParameters.authors().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("author")
+                    .getSpecification(searchParameters.authors()));
+        }
+        if (searchParameters.isbn() != null && searchParameters.isbn().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("isbn")
+                    .getSpecification(searchParameters.isbn()));
+        }
+        if (searchParameters.minPrice() != null || searchParameters.maxPrice() != null) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("price")
+                    .getSpecification(new BigDecimal[]{searchParameters.minPrice(),
+                            searchParameters.maxPrice()}));
+        }
+        if (searchParameters.descriptionKeyword() != null
+                && searchParameters.descriptionKeyword().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("description")
+                    .getSpecification(searchParameters.descriptionKeyword()));
+        }
+        return spec;
+    }
```

Comment:
This method scales very poorly, with each new parameter you have to add another block `if`. Try to modify this method in such a way that with a new filtering parameter you only need to add an item to the map, for example like this:

```java
@Override
    public Specification<Book> build(BookSearchParametersDto searchParametersDto) {
        Specification<Book> specification = Specification.where(null);
        Map<String, String[]> parametersMap = new HashMap<>();
        parametersMap.put("title", searchParametersDto.titles());
        parametersMap.put("author", searchParametersDto.authors());
        parametersMap.put("price",
                new String[]{searchParametersDto.minPrice(), searchParametersDto.maxPrice()});

        for (Map.Entry<String, String[]> entry : parametersMap.entrySet()) {
            String[] paramValues = entry.getValue();
            if (paramValues != null && paramValues.length > 0) {
                specification = specification.and(
                        specificationProviderManager.getSpecificationProvider(entry.getKey())
                                .getSpecification(paramValues));
            }
        }
        return specification;
    }
```

Reply:
I had issues with converting  the min and max price in to the string parametersMap I tried different ways, code would compile but method wouldnt work properly keep giving me nullPointerExceptions I decided to add it as a sepperate if straight in to spec